pkg/autodiscovery/listeners/snmp.go

```diff
@@ -0,0 +1,351 @@
+	adIdentifier string
+	entityID     string
+	deviceIP     string
+	creationTime integration.CreationTime
+}
+
+// Make sure SNMPService implements the Service interface
+var _ Service = &SNMPService{}
+
+type snmpSubnet struct {
+	adIdentifier  string
+	config        util.SNMPConfig
+	defaultParams *gosnmp.GoSNMP
+	startingIP    net.IP
+	currentIP     net.IP
+	network       net.IPNet
+	cacheKey      string
+}
+
+// NewSNMPListener creates a SNMPListener
+func NewSNMPListener() (ServiceListener, error) {
+	var snmpConfig util.SNMPListenerConfig
+	err := config.Datadog.UnmarshalKey("snmp_listener", &snmpConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &SNMPListener{
+		services: map[string]Service{},
+		devices:  map[string][]string{},
+		stop:     make(chan bool),
+		config:   snmpConfig,
+	}, nil
+}
+
+// Listen periodically refreshes devices
+func (l *SNMPListener) Listen(newSvc chan<- Service, delSvc chan<- Service) {
+	// setup the I/O channels
+	l.newService = newSvc
+	l.delService = delSvc
+
+	go l.checkDevices()
+}
+
+func (l *SNMPListener) loadCache(config util.SNMPConfig, adIdentifier string, cacheKey string) {
+	cacheValue, err := persistentcache.Read(cacheKey)
+	if err != nil {
+		log.Errorf("Couldn't read cache for %s: %s", cacheKey, err)
+		return
+	}
+	if cacheValue == "" {
+		return
+	}
+	var devices []net.IP
+	err = json.Unmarshal([]byte(cacheValue), &devices)
+	if err != nil {
+		log.Errorf("Couldn't unmarshal cache for %s: %s", cacheKey, err)
+		return
+	}
+	for _, deviceIP := range devices {
+		entityID := config.Digest(deviceIP.String())
+		l.createService(deviceIP.String(), adIdentifier, entityID)
+	}
+}
+
+func writeCache(cacheKey string, devices []string) {
+	cacheValue, err := json.Marshal(devices)
+	if err != nil {
+		log.Errorf("Couldn't marshal cache: %s", err)
+		return
+	}
+	err = persistentcache.Write(cacheKey, string(cacheValue))
+	if err != nil {
+		log.Errorf("Couldn't write cache: %s", err)
+	}
+}
+
+// Don't make it a method, to be overridden in tests
+var worker = func(l *SNMPListener, jobs <-chan snmpSubnet) {
+	for {
+		select {
+		case <-l.stop:
+			log.Debug("Stopping SNMP worker")
+			return
+		case subnet := <-jobs:
+			log.Debugf("Handling IP %s", subnet.currentIP.String())
+			l.checkDevice(subnet.adIdentifier, subnet.currentIP.String(), subnet.config, subnet.defaultParams)
+		default:
+		}
+	}
+}
+
+func (l *SNMPListener) checkDevice(adIdentifier string, deviceIP string, config util.SNMPConfig, defaultParams *gosnmp.GoSNMP) {
+	params := *defaultParams
+	params.Target = deviceIP
+	entityID := config.Digest(deviceIP)
+	err := params.Connect()
+	if err != nil {
+		log.Debugf("SNMP connect to %s error: %v", deviceIP, err)
+		l.deleteService(entityID)
+	} else {
+		defer params.Conn.Close()
+
+		oids := []string{"1.3.6.1.2.1.1.2.0"}
+		v, err := params.Get(oids)
+		if err != nil || len(v.Variables) < 1 || v.Variables[0].Value == nil {
+			log.Debugf("SNMP get to %s error: %v", deviceIP, err)
```

Comment:
can we handle each case with the corresponding description in the logs?
the current logic will log something like this `SNMP get to <ip> error: nil` when `err == nil` but `len(v.Variables) < 1 || v.Variables[0].Value == nil`